File: leetcode/src/01770_maximum_score_from_performing_multiplication_operations/solution.go
package leetcode

import "math"

func maximumScore(nums []int, multipliers []int) int {
	m := len(multipliers)
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, m)
		for j := range dp[i] {
			dp[i][j] = math.MinInt
		}
	}
	return solve(nums, multipliers, 0, 0, dp)
}

func solve(nums []int, mulitpliers []int, op int, left int, dp [][]int) int {
	if op == len(mulitpliers) {
		return 0
	}
	if dp[op][left] != math.MinInt {
		return dp[op][left]
	}
	dp[op][left] = max(
		mulitpliers[op]*nums[left]+solve(nums, mulitpliers, op+1, left+1, dp),
		mulitpliers[op]*nums[len(nums)-1-op+left]+solve(nums, mulitpliers, op+1, left, dp),
	)
	return dp[op][left]
}

func max(x int, y int) int {
	if x < y {
		return y
	}
	return x
}

File: leetcode/src/01770_maximum_score_from_performing_multiplication_operations/solution.rs
struct Solution;

impl Solution {
    pub fn maximum_score(nums: Vec<i32>, multipliers: Vec<i32>) -> i32 {
        let n = nums.len();
        let m = multipliers.len();
        let mut dp = vec![0; m + 1];
        for op in (0..m).rev() {
            for left in 0..=op {
                dp[left] = (multipliers[op] * nums[left] + dp[left + 1])
                    .max(multipliers[op] * nums[n - 1 - op + left] + dp[left]);
            }
        }
        dp[0]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let nums = vec![1, 2, 3];
        let multipliers = vec![3, 2, 1];
        assert_eq!(Solution::maximum_score(nums, multipliers), 14);

        let nums = vec![-5, -3, -3, -2, 7, 1];
        let multipliers = vec![-10, -5, 3, 4, 6];
        assert_eq!(Solution::maximum_score(nums, multipliers), 102);
    }
}
